Add a section about Migrating from monolithic to family Official Providers (#174)
upbound/docs: content/providers/migration.md

For manual migration to the family Official Providers refer to the
[migration guide]({{<ref "/knowledge-base/migrate-to-provider-families">}}).

`family-migrator` is a tooling for the automated migration process. The tool
supports interactive and non-interactive migration to the family Official
Providers. The instructions on how to use tooling are below.

> The `family-migrator` tool has two sub-commands, `generate` and `execute`.

Author: Added an Installation and Pre-requisites section to the document.

<!-- vale alex.ProfanityUnlikely = NO -->
<!-- ignore "executed" -->
- `generate`: For generating the migration plan. After the tool creates the

> migration plan, asks if the user wants to continue to perform the created plan.

Author: Done!

- `execute`: For only performing the generated plan. To run this command,
you must have a generated plan. Users can use this option for a generated but
un-executed plan.

> <!-- vale alex.ProfanityUnlikely = YES -->

Author: The migration tool generates a plan based on the resources available in the file system and cluster. For example, a migration plan can be generated without any MR in the cluster or a configuration package in the File System. Therefore, I think it might be confusing to add something here, as the situation may differ in each customer environment.

> <!-- vale alex.ProfanityUnlikely = YES -->

Reviewer: We need to look to document the scenario we support up front.

The `family-migrator` tool needs several inputs for generating and executing the

> plan. It prompts all required options to the user after it runs.

Author: Done!

### Collecting inputs

After running the `family-migrator generate`, tool asks a series of questions

> to user:

Author: Done!

1. `? Please specify the path for the migration plan`: The answer represents the
output path of migration plan. Users need to provide a path for generating the

> plan and patch files.

Author: We do not have a default value. However, I added an example here.

2. `? Please provide the registry and organization for the provider family
packages`: The user specifies the family provider registry and organization to
migrate. The format must be `<registry host>/<organization>`,

> like: _xpkg.upbound.io/upbound_.

Author: Done!

3. `? Please select the providers that will be migrated`:
The tool needs the versions of the providers that it wants to migrate. For this
reason, the user must first select the monolithic providers she wishes to

> migrate. The tool supports multiple selections.

Author: Done!

4. `? Please specify the version of the provider-xxx family`: The user must
provide the family versions for every selected provider.
If you want to migrate to the newest version, please see the [Upbound

> Marketplace](https://marketplace.upbound.io/) for the latest version.

Author: Done!

5. `? Please enter the URL of the migration source Configuration package`: The
user must provide the URL of the currently installed Configuration Package. The
user can take the value from the `spec.package` path of

> `Configuration.pkg.crossplane.io/v1` resource.

Author: The migration tool generates a plan based on the resources available in the file system and cluster. Therefore, a plan suitable for this situation will be generated if there is no configuration package. I added a statement that they could skip answering the question for such a case.

> `Configuration.pkg.crossplane.io/v1` resource.

Reviewer: Can we add this assumption/requirement somewhere at the top so the end user knows whether the tool can work for their scenario?

6. `? Please enter the URL of the migration target Configuration package`: The
tool builds a new configuration package during migration. Specify the URL to

> which the user pushes the built package.

Author: If a configuration package exists in the cluster, the newly generated package must be built and pushed. Because this new package URL will now be referenced in the configuration package in the cluster. Therefore, the configuration package resource will only be ready if the newly built package is pushed.

However, this step may not seem necessary because the cluster has no configuration package resource. On the other hand, the user can skip answering this and Skip the push step while running Step-by-Step.

> which the user pushes the built package.

Reviewer: That's fair enough. Maybe we can add a hint in the docs to allude to this?

7. `? Please specify the source directory for the Crossplane Configuration
package`: The user must specify the source directory where the configuration
package located. This directory contains the Configuration metadata,
compositions, and similar configuration package contents. Example:

> _/Users/user/workspace/platform-ref-gcp/package_

Author: Done!

8. `? Please specify the path to the directory containing the Crossplane package
examples`: The user must specify the directory where the configuration package

> examples located. Example: _/Users/user/workspace/platform-ref-gcp/examples_

Author: As mentioned in other answers, if there is no instance in the FileSystem (this is usually a file containing claims), the work is done by reading the claims in the cluster. Currently, the migration tool is reading resources from two sources simultaneously.

9. `? Please specify the path to store the updated configuration package`: It's

> the path where the newly built configuration package saved.

Author: Done!

<!-- vale alex.ProfanityUnlikely = NO -->
<!-- ignore "execution" -->
After the tool collects all the inputs it needs, it generates and exports the
migration plan to the specified path. Then, it asks if the user wants to

> continue to perform the created plan.

Author: Done!

<!-- vale alex.ProfanityUnlikely = YES -->

{{<hint "note" >}}
Generating the plan may take some time, depending on the number of providers and
CRDs in your cluster. Please do not stop the process until it completes. After
generation, the tool shows a message to the user that the plan has been
generated.
{{< /hint >}}

### Review generated plan, manifests, and patch files
<!-- vale alex.ProfanityUnlikely = NO -->
<!-- ignore "execution" -->
After the user chooses to move on to the execution phase, they're first
asked to review the generated plan and manifests. After a validation question
that the user has reviewed the plan and files, the tool asks whether to list the
instructions perform.

The primary purpose of the review is to make sure there are no setbacks
in the plan. In the following step, execution starts. The tool performs the
instructions generated in the plan (and displayed on the screen if the user

> chooses yes) step by step.

Author: Done!

{{<hint "note" >}}
The review process is optional and does not directly affect the transition to
the execution process. However, it is strongly recommended to complete the
review process.
{{< /hint >}}

### Execution

After collecting the inputs and reviewing the generated plan, there is no

> obstacle to proceed to the execution phase. The tool has two mods for execution:

Author: Done!

<!-- vale alex.ProfanityUnlikely = YES -->

`? Do you want to execute the migration plan with step-by-step confirmation or
no interaction`

- **Step-by-Step (Interactive):** In this option, the tool runs each step with
the consent of the user, and the user decides how to perform the step.
- **No Interaction:** In this option, the tool runs all steps end-to-end without
any interaction.

<!-- vale alex.ProfanityUnlikely = NO -->
<!-- ignore "execution" -->
#### Step-by-Step (interactive) execution

If the user chooses the Step-by-Step execution, then for every step, the tool

> wants consent and execution option. Three execution options:

Author: Done!

`? Step (with name "backup-managed-resources" at index 0) to execute:
sh -c "kubectl get managed -o yaml > backup/managed-resources.yaml"
What is your execution preference?`
<!-- vale alex.ProfanityUnlikely = YES -->

- **Automatically**: The tool performs the step automatically and shows
  the output.
- **Manually**: The tool doesn't perform the command and prompts the user for
confirmation whether run the command.
- **Skip**: The tool skips this step. Please be careful while choosing the Skip
option because you can't return to the skipped step.

<!-- vale alex.ProfanityUnlikely = NO -->
<!-- ignore "execution" -->
{{<hint "important" >}}
The Automatically option is recommended. The kubectl tool is used in the
commands to be run, and it is assumed that this tool is installed on the device
during Automatically run. Before starting the execution phase, please ensure

> kubectl is installed on your device.

Author: Done!

{{< /hint >}}
<!-- vale alex.ProfanityUnlikely = YES -->

{{<hint "note" >}}
Steps of backup the resources may take time, depending on the number of

> resources, installed providers, and CRDs in the cluster.

Author: Done!

{{< /hint >}}

<!-- vale alex.ProfanityUnlikely = NO -->
<!-- ignore "execution" -->
#### Non-Interactive execution
<!-- vale alex.ProfanityUnlikely = YES -->

> The user makes no selections or actions while running the steps in this option.
> The tool runs each step, informs the user with information-level
> messages.

Reviewer: ```suggestion
In this mode, the user takes no action and the tool runs each step, and 
informs the user of the progress.
```

Author: Done!

<!-- vale alex.ProfanityUnlikely = NO -->
<!-- ignore "execution" -->
{{<hint "important" >}}
The kubectl tool is used in the commands to be run, and it is assumed that this
tool is installed on the device during non-interactive run. Before starting the
execution phase, please ensure kubectl is installed on your device.
{{< /hint >}}

> 

Author: Moved!

If the tool doesn't report any error during the execution phase, it means that
the migration tool has worked well. The user can verify the correct

> completion of the migration by checking the status of the cluster.

Reviewer: Can we add the commands on how to check the status?

> completion of the migration by checking the status of the cluster.

Author: Done!

For manual migration to the family Official Providers refer to the
[migration guide]({{<ref "/knowledge-base/migrate-to-provider-families">}}).

`family-migrator` is a tool to automate the migration process from the
monolithic to provider family architecture. The tool supports an interactive and
non-interactive migration path.

### Pre-requisites

- `kubectl`: Before starting the execution phase, please ensure you have the
`kubectl` tool installed. `family-migrator` uses the `kubectl` tool in the
execution phase, and it assumes availability of this tool locally during the
non-interactive run.
<!-- vale Upbound.Spelling = NO -->
<!-- vale gitlab.SubstitutionWarning = NO -->
<!-- ignore "kubeconfig" -->
- The user needs to be on the device that has access to the Kubernetes cluster.
If the default config file `(~/.kube/config)` doesn't refer to the cluster to
migrate, the user must configure the `--kubeconfig` flag of the tool. The value

> of the `--kubeconfig` flag must refer to the cluster to migrate.

Author: Done!

<!-- vale Upbound.Spelling = YES -->
<!-- vale gitlab.SubstitutionWarning = YES -->
- The `family-migrator` tool currently supports a limited scenario. The relevant
scenario has some prerequisites:
  - The user must have an installed configuration package in cluster. In other
  words, the cluster must has a `Configuration.pkg.crossplane.io/v1` resource
  and the monolithic providers.
  - The user's local file system must have a configuration package and
  examples.
    - Content of Package: This directory contains the
    Configuration.meta.pkg.crossplane.io manifest (that contains the provider
    dependencies), Compositions, and CompositeResourceDefinitions. Example
    configuration package:
    https://github.com/upbound/platform-ref-gcp/tree/main/package.
    - Content of Examples: This directory may contain Claims, and
    Configuration.pkg.crossplane.io resource. Example configuration examples:
    https://github.com/upbound/platform-ref-gcp/tree/main/package.
  - To push the new configuration package image during migration, the user

>   requires push authorization to the registry.

Author: Done!

### Installation

1. Please use the [Releases] page to download the `family-migrator` tool. Download
the appropriate binary for your operating system (`Linux` or `Darwin/macOS`) and
architecture (`amd64` or `arm64`). The download is an executable file
which might require the necessary file permissions to allow execution.
2. Move the downloaded binary to `/usr/local/bin`
3. Give permission to execute the binary
```shell
chmod +x family-migrator_darwin-arm64
```

{{<hint "note" >}}
In Darwin/macOS, the following error can be shown while trying to run the
executable. Please allow this application in the Security Settings by using the
`Allow Anyway`.
{{< /hint >}}
{{<img src="providers/images/malicious-software.jpeg" alt="Malicious Software Error" size="tiny">}}

> {{<img src="providers/images/allow-anyway.jpeg" alt="Allow Anyway" size="small">}}

Author: Done!

### Usage

The `family-migrator` tool has two sub-commands, `generate` and `execute`.

The help output of the `family-migrator`:

{{<img src="providers/images/family-migrator-help.jpeg" alt="family-migrator help output" size="medium">}}

- `generate`: This command generates the migration plan. After the tool creates
the migration plan, it confirms if execution of the created plan should proceed.

- `execute`: Used for executing a generated migration plan. You must generate a
migration plan first before running this command.

The `family-migrator` tool needs several inputs for generating and executing the
plan. It prompts for all required options when executing the command.

#### Required inputs

When running the `family-migrator generate` command, the tool prompts the
user with a series of questions to gather the required inputs:

<!-- vale Google.Will = NO -->
<!-- vale gitlab.FutureTense = NO -->
<!-- vale write-good.Passive = NO -->
{{< table "table table-sm table-striped">}}
| Required input | Description |
| ----- | ----------- |

> | **Please specify the path for the migration plan** | The answer represents the output path of the migration plan. Users need to provide a path for generating the plan and patch files. For example, `/tmp/generated/migration-plan.yaml` |

Author: Let me add this feature to the https://github.com/upbound/extensions-migration/pull/19

> | **Please specify the path for the migration plan** | The answer represents the output path of the migration plan. Users need to provide a path for generating the plan and patch files. For example, `/tmp/generated/migration-plan.yaml` |

Author: I thought that adding this quickly to the PR. However, it may be a bit changes related to variables and we need to test this carefully. So, I will open an issue and let's address it then :)

| **Please provide the registry and organization for the provider family packages** | The format must be `<registry host>/<organization>`, for example, `xpkg.upbound.io/upbound`. |
| **Please select the providers that will be migrated** | The tool presents user with a list of the monolithic providers they can select to migrate from. The tool supports multiple selections. |

> | **Please specify the version of the provider-abc family** | Provide the versions of the family provider to migrate to for each selected provider. If you want to migrate to the latest version, refer to the [Upbound Marketplace](https://marketplace.upbound.io). The version format must be `v0.y.y` |

Author: With this [PR](https://github.com/upbound/extensions-migration/pull/19) we will list some suggestions to the user. 

An example output for aws:

```bash
? Please specify the version of the provider-aws family. Possible versions:
v0.37.0
v0.36.0
```

User can be copy and paste from there now. So, If it's okay with you, we can leave it that way for now.

| **Please enter the URL of the migration source Configuration package** | Provide the registry URL of the currently installed Configuration Package. The value appears in the `spec.package` path of `Configuration.pkg.crossplane.io/v1` resource. For example, `xpkg.upbound.io/upbound/platform-ref-gcp:v0.3.0` |

> | **Please enter the URL of the migration target Configuration package** | The tool builds a new configuration package during migration. Specify the registry URL where to update the built package. For example, `<registry-host>/<organization>/platform-ref-gcp:v0.3.0-migrated` |

Author: Done!

| **Please specify the source directory for the Crossplane Configuration package** | Specify the source directory to the location of the configuration package. This directory contains the Configuration.meta.pkg.crossplane.io manifest that contains the provider dependencies, Compositions, and CompositeResourceDefinitions. For example: `/Users/user/workspace/platform-ref-gcp/package` Example configuration package: https://github.com/upbound/platform-ref-gcp/tree/main/package. |
| **Please specify the path to the directory containing the Crossplane package examples** | Specify the directory to the location of the configuration package examples. This directory may contain Claims, and Configuration.pkg.crossplane.io resource. For example: `/Users/user/workspace/platform-ref-gcp/examples` Example configuration examples: https://github.com/upbound/platform-ref-gcp/tree/main/examples. |

> | **Please specify the path to store the updated configuration package** | The location to store the newly built configuration package. |

Reviewer: Can't we just re-use the path from step 1?

Author: We started to use it in this [PR](https://github.com/upbound/extensions-migration/pull/19). I removed this.

provider. If you want to migrate to the latest version, refer to the [Upbound
Marketplace]
5. `? Please enter the URL of the migration source Configuration package`: The
user must provide the URL of the currently installed Configuration Package. The
user can take the value from the `spec.package` path of
`Configuration.pkg.crossplane.io/v1` resource.
6. `? Please enter the URL of the migration target Configuration package`: The
tool builds a new configuration package during migration. Specify the URL to
which the user pushes the built package.
7. `? Please specify the source directory for the Crossplane Configuration
package`: The user must specify the source directory where the location of the
configuration package. This directory contains the Configuration metadata,
compositions, and similar configuration package contents. Example:
`/Users/user/workspace/platform-ref-gcp/package`
8. `? Please specify the path to the directory containing the Crossplane package
examples`: The user must specify the directory where the location of the
configuration package examples. Example:
`/Users/user/workspace/platform-ref-gcp/examples`

> 9. `? Please specify the path to store the updated configuration package`: This
> is where the location of the newly built configuration package.

Reviewer: ```suggestion
9. `? Please specify the path to store the updated configuration package`: Specify
the directory to the location where the newly built configuration package should be
saved.
```

<!-- vale alex.ProfanityUnlikely = NO -->
<!-- ignore "execute" -->

> After the tool collects all the inputs it needs, it generates and exports the
> migration plan to the specified path. Then, it asks if the user wants to
> continue to execute the created plan.

Reviewer: ```suggestion
After the tool collects all the required inputs, it generates and exports the
migration plan to the specified path. It will prompt the user whether it should
continue to execute the created plan.
```

<!-- vale alex.ProfanityUnlikely = YES -->

{{<hint "note" >}}
Generating the plan may take some time, depending on the number of providers and
CRDs in your cluster. Please do not stop the process until it completes. After
generation, the tool shows a message to the user that the plan has been

> generated.

Reviewer: Is it possible to have some output message of progress to the user during this stage so that they stay informed on progress?

{{< /hint >}}

#### Review generated plan, manifests, and patch files
<!-- vale alex.ProfanityUnlikely = NO -->
<!-- ignore "execution" -->
After the user chooses to move on to the execution phase, they're first
asked to review the generated plan and manifests. After a validation question
that the user has reviewed the plan and files, the tool asks whether to list the
instructions to execute.

> The review's primary purpose is to ensure no unexpected steps are
> in the plan. In the following step, execution starts. The tool performs the
> instructions generated in the plan (and displayed on the screen if the user
> chooses yes) step by step.

Reviewer: ```suggestion
```
